src/client-side-encryption/state_machine.ts

```diff
@@ -153,176 +151,125 @@ export class StateMachine {
     private bsonOptions = pluckBSONSerializeOptions(options)
   ) {}
 
-  executeAsync<T>(executor: StateMachineExecutable, context: MongoCryptContext): Promise<T> {
-    // @ts-expect-error The callback version allows undefined for the result, but we'll never actually have an undefined result without an error.
-    return promisify(this.execute.bind(this))(executor, context);
-  }
-
   /**
    * Executes the state machine according to the specification
    */
-  execute<T extends Document>(
+  async execute<T extends Document>(
     executor: StateMachineExecutable,
-    context: MongoCryptContext,
-    callback: Callback<T>
-  ) {
+    context: MongoCryptContext
+  ): Promise<T> {
     const keyVaultNamespace = executor._keyVaultNamespace;
     const keyVaultClient = executor._keyVaultClient;
     const metaDataClient = executor._metaDataClient;
     const mongocryptdClient = executor._mongocryptdClient;
     const mongocryptdManager = executor._mongocryptdManager;
+    let result: T | null = null;
+
+    while (context.state !== MONGOCRYPT_CTX_DONE && context.state !== MONGOCRYPT_CTX_ERROR) {
```

Comment:
Was there a reason for refactoring this to use a `while` loop instead of recursion like before?

Reply:
Neal and I discussed it and wanted to do it just for clarity (I like the loop better).  But I'm also glad I did it, given Neal's memory leak investigation with `AbstractCursor.next()`'s recursive calls potentially resulting in many promise allocations.